pandas/core/generic.py

```diff
@@ -8519,12 +8520,14 @@ def _align_series(
             right = right.fillna(fill_value, method=method, limit=limit)
 
         # if DatetimeIndex have different tz, convert to UTC
-        if is_series or (not is_series and axis == 0):
-            if is_datetime64tz_dtype(left.index):
-                if left.index.tz != right.index.tz:
-                    if join_index is not None:
-                        left.index = join_index
-                        right.index = join_index
+        if (
+            (is_series or (not is_series and axis == 0))
+            and is_datetime64tz_dtype(left.index.dtype)
+            and (left.index.tz != right.index.tz)
+            and (join_index is not None)
+        ):
```

Comment:
sometimes we need to add casts or asserts for mypy after is_* calls. I would prefer to not to do this type of refactor at this time.